Index: trunk/src/exif.cpp
===================================================================
--- trunk/src/exif.cpp	(revision 2016)
+++ trunk/src/exif.cpp	(revision 2017)
@@ -601,12 +601,12 @@
 
     void ExifData::sortByKey()
     {
-        std::sort(exifMetadata_.begin(), exifMetadata_.end(), cmpMetadataByKey);
+        exifMetadata_.sort(cmpMetadataByKey);
     }
 
     void ExifData::sortByTag()
     {
-        std::sort(exifMetadata_.begin(), exifMetadata_.end(), cmpMetadataByTag);
+        exifMetadata_.sort(cmpMetadataByTag);
     }
 
     ExifData::iterator ExifData::erase(ExifData::iterator beg, ExifData::iterator end)
Index: trunk/src/exif.hpp
===================================================================
--- trunk/src/exif.hpp	(revision 2016)
+++ trunk/src/exif.hpp	(revision 2017)
@@ -38,7 +38,7 @@
 
 // + standard includes
 #include <string>
-#include <vector>
+#include <list>
 #include <memory>
 
 // *****************************************************************************
@@ -423,7 +423,7 @@
     }; // class ExifThumb
 
     //! Container type to hold all metadata
-    typedef std::vector<Exifdatum> ExifMetadata;
+    typedef std::list<Exifdatum> ExifMetadata;
 
     /*!
       @brief A container for Exif data.  This is a top-level class of the %Exiv2
Index: trunk/src/tiffcomposite.cpp
===================================================================
--- trunk/src/tiffcomposite.cpp	(revision 2016)
+++ trunk/src/tiffcomposite.cpp	(revision 2017)
@@ -161,8 +161,7 @@
 
     bool TiffGroupInfo::operator==(const std::string& groupName) const
     {
-        std::string name(name_);
-        return name == groupName;
+        return 0 == strcmp(name_, groupName.c_str());
     }
 
     const char* tiffGroupName(uint16_t group)
