Index: exiv2-0.22/src/minoltamn.cpp
===================================================================
--- exiv2-0.22.orig/src/minoltamn.cpp	2011-02-14 01:08:44.000000000 +1100
+++ exiv2-0.22/src/minoltamn.cpp	2012-01-25 06:39:28.000000000 +1100
@@ -1100,9 +1100,9 @@
         { 8,    N_("White Balance Bracketing Low")  },
         { 9,    N_("White Balance Bracketing High") },
         { 770,  N_("Single-frame Bracketing Low")   },
-        { 771,  N_("Continous Bracketing Low")      },
+        { 771,  N_("Continuous Bracketing Low")      },
         { 1794, N_("Single-frame Bracketing High")  },
-        { 1795, N_("Continous Bracketing High")     }
+        { 1795, N_("Continuous Bracketing High")     }
     };
 
     //! Lookup table to translate Sony A100 camera settings focus mode values to readable labels
Index: exiv2-0.22/src/sonymn.cpp
===================================================================
--- exiv2-0.22.orig/src/sonymn.cpp	2011-02-14 01:08:44.000000000 +1100
+++ exiv2-0.22/src/sonymn.cpp	2012-01-25 06:40:05.000000000 +1100
@@ -412,7 +412,7 @@
                 N_("Release Mode"),
                 sony1Id, makerTags, unsignedShort, -1, EXV_PRINT_TAG(sonyReleaseMode)),
         TagInfo(0xB04A, "SequenceNumber", N_("Sequence Number"),
-                N_("Shot number in continous burst mode"),
+                N_("Shot number in continuous burst mode"),
                 sony1Id, makerTags, unsignedShort, -1, EXV_PRINT_TAG(sonySequenceNumber)),
         TagInfo(0xB04B, "AntiBlur", N_("Anti-Blur"),
                 N_("Anti-Blur"),
